File: minimum-number-of-operations-to-make-string-sorted.cc
// Minimum Number of Operations to Make String Sorted
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)
#define ROF(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (b); --i >= (a); )

class Solution {
  static const int MOD = 1000000007;
  int inv(int x) {
    long s = 1;
    for (; x > 1; x = MOD%x)
      s = s*(MOD-MOD/x)%MOD;
    return s;
  }
public:
  int makeStringSorted(string s) {
    int n = s.size(), cnt[26] = {};
    vector<int> fac(n+1), invfac(n+1);
    fac[0] = invfac[0] = 1;
    FOR(i, 1, n+1) {
      fac[i] = long(fac[i-1])*i%MOD;
      invfac[i] = inv(fac[i]);
    }
    long ans = 0;
    ROF(i, 0, n) {
      int ch = s[i]-'a';
      long c = accumulate(cnt, cnt+ch, 0L)*fac[n-1-i]%MOD;
      cnt[ch]++;
      REP(j, 26)
        c = c*invfac[cnt[j]]%MOD;
      ans = (ans+c)%MOD;
    }
    return ans;
  }
};

File: minimum-number-of-operations-to-make-string-sorted.rs
// Minimum Number of Operations to Make String Sorted
const MOD: i64 = 1000000007;

impl Solution {
    fn inv(mut x: i32) -> i32 {
        let mut s = 1;
        while x > 1 {
            s = s * (MOD-MOD/x as i64) % MOD;
            x = MOD as i32 % x;
        }
        return s as i32;
    }
    pub fn make_string_sorted(s: String) -> i32 {
        let n = s.len();
        let mut fac = vec![1; n+1];
        let mut invfac = vec![1; n+1];
        for i in 1..=n {
            fac[i] = (fac[i-1] as i64 * i as i64 % MOD) as i32;
            invfac[i] = Self::inv(fac[i]);
        }
        let mut cnt: Vec<usize> = vec![0; 26];
        let mut ans = 0;
        for i in (0..n).rev() {
            let ch = (s.as_bytes()[i]-b'a') as usize;
            let mut c = cnt[0..ch].iter().sum::<usize>() as i64 * fac[n-1-i] as i64 % MOD;
            cnt[ch] += 1;
            for j in 0..26 {
                c = c * invfac[cnt[j]] as i64 % MOD;
            }
            ans = (ans+c)%MOD;
        }
        ans as i32
    }
}
